h2o-algos/src/main/java/hex/glrm/GLRM.java

```diff
@@ -1131,13 +1186,29 @@ private Chunk chk_xnew(Chunk[] chks, int c) {
 
           // Gradient wrt x_i is matrix product \grad L_{i,j}(x_i * Y_j, A_{i,j}) * Y_j'
           double[] weight = _lossFunc[j].mlgrad(xy, (int) a[j]);
-          double[][] ysub = _yt.getCatBlock(j);
+ //         double[][] ysub = _yt.getCatBlock(j);
           for (int k = 0; k < _ncolX; k++) {
-            for (int c = 0; c < weight.length; c++)
-              grad[k] += cweight * weight[c] * ysub[k][c];
+            for (int c = 0; c < weight.length; c++) {
+ //             grad[k] += cweight * weight[c] * ysub[k][c];
+              int cidx = _yt.getCatCidx(j, c);
+              double archtypevalues = 0;
+              if (_yt._transposed)
+                archtypevalues = _yt._archetypes[cidx][k];
```

Comment:
It would probably pay off to transpose the matrix physically in the beginning to optimize on memory access:
archtypevalues = _yt._archetypes[cidx][k] is really slow